### PR TITLE
Soho 6787 provider declaration in the sohodropdowncomponent

### PR DESCRIPTION
**Pull request by bthharper**
That's the problem, the model changes, but the dropdown doesn't. I'm struggling to see a way to make this work without a lot of code or some _evil_ JavaScript code.

Is there a way of detecting the value attribute being set on the select element?

This is the code in the Angular Control Value Accessor:

`this._renderer.setProperty(this._elementRef.nativeElement, 'value', valueString);`

If I could detect this, I could call `selectValue` on the dropdown.

Currently, I have added a `setTimeout` to call updated after initialization.

### CODE REVIEW

**Comment by bthharper** [comment]
I seem to have ended up with the Angular 6 changes in this too, I assume because i had to merge them into this branch as I'd already created it.  Any ideas?  @clepore

**Comment by clepore** [comment]
Hmm yeah. If you branched it from master or your old branch, and didn't rebsae it or merge the new master into it, there is a chance some commits from the revert might be kept around.

**Comment by pwpatton** [comment]
@bthharper should we wait on this PR until the ng6 PR is ready to be merged as well?

**Comment by bthharper** [comment]
It removes the NgModel, and the examples seem to work better.  I have a concern (especially due to the number of issues on teams with dropdown - specifically the angualr component) that it might cause issues elsewhere.

**Comment by bthharper** [comment]
Maybe we should push it, as we can always revert it laters.  I added some zone code in too, which you may want to review.

**Comment by pwpatton** [comment]
we use it extensively, perhaps we could make the change locally and verify it. Would these changes work against ng5 version?

**Comment by fitzorama** [comment]
I've brought soho-dropdown.component.ts and soho-dropdown.d.ts into Landmark and also to my 4.6.0-rc angular components repo.

In Landmark, I am seeing an issue when using the source property for a single select and when configured as a multi-select. When clicking the dropdown, the first click only shows the selected items and does not populate; the second click will correctly display the populated items.

In the angular components repo, I am seeing something similar for the source property and multi-select only, where only the selected items are displayed until a change is made and then the populated items are displayed thereafter. Use the dropdown-multiselect.demo with either the 'Using alpha source & tooltip' or 'Using numeric source' fields to see this: Change any dropdown and then the source populates on those fields.

File: src/soho/dropdown/soho-dropdown.component.ts [hunk 1]
```diff
@@ -312,35 +319,78 @@ export class SohoDropDownComponent implements AfterViewInit, OnDestroy {
...
+
+      // @todo - add event binding control so we don't bind if not required.
+      this.jQueryElement
+        .on('change', (event: JQuery.Event) => this.onChange(event))
+        .on('updated', (event: JQuery.Event) => this.updatedEvent.emit(event));
+
+      setTimeout(() => {
+        this.updated();
+      }, 1);
+    });
```
[the remark this answers is not in the document]
Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]: The FormModel has not been updated when that lifecycle event is called, so calling update does not recognise the value.  Putting the timeout pushed it passed angular initialising the model.  In an ideal world there would be a callback on the component AfterModelInit, but alas not.

File: src/soho/dropdown/soho-dropdown.component.ts [hunk 2]
```diff
@@ -312,35 +319,78 @@ export class SohoDropDownComponent implements AfterViewInit, OnDestroy {
...
+      // call outside the angular zone so change detection isn't
+      // triggered by the soho component.
+      this.ngZone.runOutsideAngular(() => {
+        // remove the event listeners on this element.
+        this.jQueryElement.off();
+
+        // Destroy any widget resources.
+        this.dropdown.destroy();
+        this.dropdown = null;
+      });
```
[the remark this answers is not in the document]
Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]: Good call ... thought I had done that.

File: src/soho/dropdown/soho-dropdown.component.ts [hunk 3]
```diff
@@ -389,11 +444,27 @@ export class SohoDropDownComponent implements AfterViewInit, OnDestroy {
...
+   *
+   * Sets the value of the dropdown.
+   *
+   * This is the model value that is to be set.
+   * @param value
+   */
+  public selectValue(value: any): void {
+    if (this.dropdown) {
+      // @todo lookup model value!
+      this.dropdown.selectValue(value);
```
[the remark this answers is not in the document]
Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]: Good spot.

**Comment by bthharper** [comment]
Yes - that's a problem I am seeing too.  Are you calling `dropdown.updated()` in your code in the `source` callback?

File: src/soho/dropdown/soho-dropdown.component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pwpatton) [reply src/soho/dropdown/soho-dropdown.component.ts]: I've been using your `ngAfterViewChecked()` technique from datagrid for calling `update()` if it's necessary in the next tick. Something like this:
```
ngAfterViewInit() {
   ... soho initialization stuff
   this.runUpdatedOnCheck = true; // or something like that.
}
ngAfterViewChecked() {
   if (this.runUpdatedOnCheck) {
      this.updated();
      this.runUpdatedOnCheck = false;
   }
}
```
I realize that `ngAfterViewInit()` is just the first call of `ngAfterViewChecked` so I don't know if changeDetection to cause another `ngAfterViewChecked` call. If so (which seem likely) you'd have to place the runUpdatedOnCheck in a timeout anyway. Probably not worth it unless you are calling updated from a number of places that require it to be called in the next tick (i.e. after the dom is updated from angular).

File: src/soho/dropdown/soho-dropdown.component.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]: @pwpatton are you sure about the `NgZone.assertNotInAngular()` ... won't this method be called inside the angular zone or am I misinterpreting your comment?

[the remark this answers is not in the document]
Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]:
```typescript
public selectValue(value: any): void {
    NgZone.assertNotInAngular()
    if (this.dropdown) {
      this.ngZone.runOutsideAngular(() => {
        this.dropdown.selectValue(value);
      });
    }
  }

File: src/soho/dropdown/soho-dropdown.component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]: Hey - forgot about that let me try.

**Comment by fitzorama** [comment]
No, we are not calling `dropdown.updated()`. I did add the call in both Landmark and the demo and did not see any improvement.

**Comment by fitzorama** [comment]
@bthharper I have your soho-dropdown changes working with both Landmark and the dropdown-multiselect.demo using the changes below. 

There were two issues in the dropdown-multiselect.demo: The first was the Component decorator is using `ChangeDetectionStrategy.OnPush` and the second is the sourceAlpha() and sourceNumeric() methods are running outside of Angular. 

We need to run markForCheck() on the changeDetector, but because the source methods are outside of Angular, the setTimeout is not triggering change detection.

Adding the following to the sourceAlpha() and sourceNumeric() methods resolved the issues and the dropdowns display fully populated on the first click.

```
sourceAlpha = 
    ... 
    this.model.multiSourceAlpha.options = options;

    this.changeDetectorRef.markForCheck();

    NgZone.assertNotInAngularZone();

    // ensure we are back in a zone so that the timeout will trigger change detection.
    this.ngZone.run(() => setTimeout(() => {
      // indicate to sohoxi dropdown to not create option tags
      response(options, true);
    }, 1));
}
```
Doing this is probably a breaking change as all callers would have to change their source call. 

Is there a way to delegate the source method so when it runs, it runs in the Angular zone? Any ideas?

File: src/soho/dropdown/soho-dropdown.component.ts [hunk 4]
```diff
@@ -325,23 +332,26 @@ export class SohoDropDownComponent implements AfterViewInit, AfterViewChecked, O
...
 
   /**
-   * @description
-   * Constructor.
-   *
-   * @param element the associated element.
-   * @param model the model
-   * @param ngZone the angular zone
+   * Creates an instance of SohoDropDownComponent.
+   * @param {ElementRef} element the element this component encapsultes.
```
Comment (fitzorama) [inline src/soho/dropdown/soho-dropdown.component.ts]: Spelling of encapsultes

File: src/soho/dropdown/soho-dropdown.component.ts [hunk 5]
```diff
@@ -487,28 +531,60 @@ export class SohoDropDownComponent implements AfterViewInit, AfterViewChecked, O
...
+/**
+ * Provides a 'wrapper' around the {ControlValueAccessor} added by
+ * angular when handling `select` elements.
+ *
+ * This class allows the {SohoDropDownComponent} to interoperate with
+ * the {ControlValueAccessor}.  Specifically, providing access to the
+ * onChange function, which we must call when the value of the dropdown
+ * is modified.
+ *
+ * @class IdsDropDownControlValueAccessor
```
Comment (fitzorama) [inline src/soho/dropdown/soho-dropdown.component.ts]: Should be SohoDropDownControlValueAccessorDelegator?

File: src/soho/dropdown/soho-dropdown.component.ts [hunk 6]
```diff
@@ -487,28 +531,60 @@ export class SohoDropDownComponent implements AfterViewInit, AfterViewChecked, O
...
+   * @memberof SohoDropDownControlValueAccessorDelegate
+   */
+  constructor(
+    private delegate: ControlValueAccessor,
+    private dropdown: SohoDropDownComponent) { }
 
-  constructor(private delegate: ControlValueAccessor,
-  private dropdown: SohoDropDownComponent) {}
+  writeValue(value: any): void {
+    console.log(`writeValue ${JSON.stringify(value)}`);
```
Comment (fitzorama) [inline src/soho/dropdown/soho-dropdown.component.ts]: Do we need this for production?

File: src/soho/dropdown/soho-dropdown.component.ts [hunk 7]
```diff
@@ -487,28 +531,60 @@ export class SohoDropDownComponent implements AfterViewInit, AfterViewChecked, O
...
+    this.delegate.writeValue(value);
 
-  writeValue(obj: any): void {
-    //this.dropdown.selectValue(obj);
-    this.delegate.writeValue(obj);
+    // Get the actual value back from the
+    // delegate.
+    this.value = (this.delegate as any).value;
+    console.log(`${JSON.stringify(this.value)}`);
```
Comment (fitzorama) [inline src/soho/dropdown/soho-dropdown.component.ts]: Do we need this for production?

Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]: Sorry - didn't realise this got pushed into the PR - this was my local testing.  I think I am going to cancel this PR and create a new one.

File: src/soho/dropdown/soho-dropdown.component.ts (hunk 5 above)
Comment (bthharper) [reply src/soho/dropdown/soho-dropdown.component.ts]: Yes - local version has that ;-)

**Comment by bthharper** [comment]
Closed this PR - I am going to create a new one with hopefully just the changes for this fix.

**Pull request closed without merging**